pkg/daemon/daemon.go

```diff
@@ -251,145 +339,76 @@ func (dn *Daemon) operatorConfigChangeHandler(old, new interface{}) {
-		dn.drainNode(nodeState.GetName())
-	}
-	for k, p := range dn.LoadedPlugins {
-		if k != GenericPlugin {
-			err := p.Apply()
-			if err != nil {
-				glog.Errorf("nodeStateAddHandler(): plugin %s fail to apply: %v", k, err)
-				dn.exitCh <- err
-				return
-			}
-		}
-	}
-
-	if len(dn.LoadedPlugins) > 1 && !reqReboot {
-		// Apply generic_plugin last
-		err = dn.LoadedPlugins[GenericPlugin].Apply()
+func (dn *Daemon) nodeStateSyncHandler(oldState, newState *sriovnetworkv1.SriovNetworkNodeState) error {
+	var err, lastErr error
+	// load plugins if has not loaded
+	if len(dn.LoadedPlugins) == 0 {
+		err = dn.loadVendorPlugins(newState)
 		if err != nil {
-			glog.Errorf("nodeStateAddHandler(): generic_plugin fail to apply: %v", err)
-			dn.exitCh <- err
-			return
-		}
-	}
-
-	if reqReboot {
-		glog.Info("nodeStateAddHandler(): reboot node")
-		rebootNode()
-		return
-	}
-
-	// restart device plugin pod
-	if reqDrain {
-		if err := dn.restartDevicePluginPod(); err != nil {
-			glog.Errorf("nodeStateAddHandler(): fail to restart device plugin pod: %v", err)
-			dn.exitCh <- err
-			return
+			glog.Errorf("nodeStateSyncHandler(): failed to load vendor plugin: %v", err)
+			return err
 		}
 	}
 
-	dn.completeDrain()
-	dn.refreshCh <- Message{
-		syncStatus:    "Succeeded",
-		lastSyncError: "",
-	}
-}
-
-func (dn *Daemon) nodeStateChangeHandler(old, new interface{}) {
-	var err, lastErr error
-	newState := new.(*sriovnetworkv1.SriovNetworkNodeState)
-	oldState := old.(*sriovnetworkv1.SriovNetworkNodeState)
-	glog.V(2).Infof("nodeStateChangeHandler(): new generation is %d", newState.GetObjectMeta().GetGeneration())
+	glog.V(0).Infof("nodeStateSyncHandler(): new generation is %d", newState.GetObjectMeta().GetGeneration())
 	// Get the latest NodeState
 	var latestState *sriovnetworkv1.SriovNetworkNodeState
 	err = wait.PollImmediate(10*time.Second, 1*time.Minute, func() (bool, error) {
 		latestState, lastErr = dn.client.SriovnetworkV1().SriovNetworkNodeStates(namespace).Get(context.Background(), dn.name, metav1.GetOptions{})
 		if lastErr == nil {
 			return true, nil
 		}
-		glog.Warningf("nodeStateChangeHandler(): Failed to fetch node state %s (%v); retrying...", dn.name, lastErr)
+		glog.Warningf("nodeStateSyncHandler(): Failed to fetch node state %s (%v); retrying...", dn.name, lastErr)
 		return false, nil
 	})
 	if err != nil {
 		if err == wait.ErrWaitTimeout {
-			dn.exitCh <- goerr.Wrapf(lastErr, "nodeStateChangeHandler(): Timed out trying to fetch the latest node state %s", dn.name)
-			return
+			return goerr.Wrapf(lastErr, "nodeStateSyncHandler(): Timed out trying to fetch the latest node state %s", dn.name)
 		}
-		dn.exitCh <- err
-		return
+		return err
 	}
-	latest := latestState.GetObjectMeta().GetGeneration()
-	if latest != newState.GetObjectMeta().GetGeneration() {
-		glog.Errorf("nodeStateChangeHandler(): the lastet generation is %d, skip", latest)
-		return
+	latest := latestState.GetGeneration()
```

Comment:
Not sure, but could this logic be moved directly into the handler, before putting the event in the queue?

Reply:
Yes, it can be done. I did consider it. But a request may be put back into the queue when 1) the node state changes 2) an error happens in the handler. So if we choose to move the logic, then we need to put it to both places, which I tend to not.

Reply:
So, this made me think to another issue.
If you put the event back on the queue for some error, you risk to have an older event to be consumed AFTER a more recent one. This is because in this patch we are putting the event itself and not just the event key (normally name + ns). 

The established pattern is to put the key of the event in the queue, and fetch the latest element from the informer cache when the key is read from the queue. This avoids the issue mentioned above (and probably many others), because you always consume the freshest.

In this case it does not come for free, because there are a few places where we use the old version (but not as many as I thought):
- The version check, that can be moved into the handler
- The `needDrain` function in generic plugin, that could check directly the value from sysfs
- The version of the DP configmap. This is probably the trickier one, which requires a bit more thought, because there's no easy way to understand if the running dp is using the current configmap or an old one. This also means that if we loose that change of dp configmap version event, we will never restart the dp.

I am afraid that relying on the old / new couple in the exit of the workqueue will leave us with new corner cases.

Reply:
> The needDrain function in generic plugin, that could check directly the value from sysfs

Yes, comparing with sysfs is more reliable. But I think we can do it in a separate PR.

> relying on the old / new couple in the exit of the workqueue will leave us with new corner cases.

I agree. And I made some changes in the latest patch to reduce the dependency on the old object. So I think from the plugins' perspective, we don't need it anymore. I think I have to solution to solve the DP config problem. Let me refactor this patch a bit.

/hold